### PR TITLE
Add validation method for RunCompletionEvent

### PR DESCRIPTION
**Pull request by edinekj**
# Summary:
Validator library added - https://pkg.go.dev/github.com/go-playground/validator/v10@v10.22.0

Added a validation extension method for `RunCompletionEvent` to validate the following fields:
- Status
- PipelineName
- RunId
- Either RunConfigurationName or RunName
- Provider 

Note: `RunConfigurationName` and `RunName` can have an empty Namespace field.

Note: Refactored the existing `suite_unit_test` to extract existing tests into own unit_test files.  So `common_types_unit_test.go` is a cut and paste from that file.

# Reason:
The Go JSON unmarshaller isn't strict, and any missing or extra fields are simply ignored. There is no way to add required fields using the Go JSON library.

### CODE REVIEW

File: argo/common/run_completion.go [hunk 1]
```diff
@@ -20,18 +25,52 @@ var RunCompletionStatuses = struct {
...
 func (sre RunCompletionEvent) String() string {
 	return fmt.Sprintf("{Status:%s PipelineName:%+v RunConfigurationName:%+v RunName:%+v RunId:%s ServingModelArtifacts:%+v Artifacts:%+v Provider:%s}",
 		sre.Status, sre.PipelineName, sre.RunConfigurationName, sre.RunName, sre.RunId, sre.ServingModelArtifacts, sre.Artifacts, sre.Provider)
 }
+
+func validateNamespacedName(nn *NamespacedName, key string) error {
+	if nn == nil {
+		return fmt.Errorf("key: %s is nil", key)
+	} else if nn.Name == "" {
+		return fmt.Errorf("key: %s, Name filed is missing", key)
```
Comment (peter-hazell) [inline argo/common/run_completion.go:48]:
```suggestion
		return fmt.Errorf("key: %s, Name field is missing", key)
```

**Review by TobyPinfold — APPROVED** [review]

File: argo/common/run_completion.go (hunk 1 above)
Comment (peter-hazell) [inline argo/common/run_completion.go:48]: Looks like there is still a typo??

Comment (edinekj) [reply argo/common/run_completion.go]: No idea how it has been missed even if we commit suggestions.

Comment (edinekj) [reply argo/common/run_completion.go]: 328d1937212e144ecb6c10f47f5f81b38edfd5c3

File: argo/common/run_completion.go [hunk 2]
```diff
@@ -20,18 +25,52 @@ var RunCompletionStatuses = struct {
...
-	Status       RunCompletionStatus `json:"status"`
-	PipelineName NamespacedName      `json:"pipelineName"`
+	Status       RunCompletionStatus `json:"status" validate:"required"`
+	PipelineName NamespacedName      `json:"pipelineName" validate:"pipelineName"`
 	// Optionally render structs until https://github.com/golang/go/issues/11939 is addressed
-	RunConfigurationName  *NamespacedName `json:"runConfigurationName,omitempty"`
-	RunName               *NamespacedName `json:"runName,omitempty"`
-	RunId                 string          `json:"runId"`
+	RunConfigurationName  *NamespacedName `json:"runConfigurationName"`
+	RunName               *NamespacedName `json:"runName"`
```
Comment (grahamia) [inline argo/common/run_completion.go:32]: These should still be omitempty.

The validation here should be exactly one should be populated (with at least a name (validate both can be a separate card in future)).

if both populated - error
if neither populated - error
if one populated then all good.

File: argo/common/run_completion.go [hunk 3]
```diff
@@ -20,18 +25,52 @@ var RunCompletionStatuses = struct {
...
+
+	return nil
+}
+
+func (sre RunCompletionEvent) Validate() error {
+	validate := validator.New()
+	validate.RegisterValidation("pipelineName", pipelineNameValidator)
+	validateErr := validate.Struct(sre)
+
+	runConfigurationNameValidationErr := validateNamespacedName(sre.RunConfigurationName, "RunCompletionEvent.RunConfigurationName")
```
Comment (grahamia) [inline argo/common/run_completion.go:59]: this needs changing following the previous comment

File: argo/common/suite_unit_test.go [hunk 4]
```diff
@@ -5,9 +5,10 @@ package common
...
 	"encoding/json"
 	"fmt"
+	"testing"
+
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
-	"testing"
 )
 
 func TestCommonUnitSuite(t *testing.T) {
```
Comment (grahamia) [inline argo/common/suite_unit_test.go:14]: This test suite doesn't follow the same pattern as all the others, i.e. having just the suite setup in this suite file and then the tests for each component in own unit_test files. So split the existing namespacename tests out into own file and create a run_completion_event_unit_test file

**Review by grahamia — CHANGES_REQUESTED** [review]

File: argo/common/run_completion.go (hunk 2 above)
Comment (TobyPinfold) [reply argo/common/run_completion.go]: changed in [f7076ff](https://github.com/sky-uk/kfp-operator/pull/352/commits/f7076ffe6e057088e9725fc23b568083c3e76cab)

File: argo/common/run_completion.go (hunk 3 above)
Comment (TobyPinfold) [reply argo/common/run_completion.go:58]: changed in [f7076ff](https://github.com/sky-uk/kfp-operator/pull/352/commits/f7076ffe6e057088e9725fc23b568083c3e76cab)

File: argo/common/suite_unit_test.go (hunk 4 above)
Comment (TobyPinfold) [reply argo/common/suite_unit_test.go]: done in [6996d38](https://github.com/sky-uk/kfp-operator/pull/352/commits/6996d3803367bab1833668a9763c2db01a8fcc73)

File: argo/common/run_completion.go [hunk 5]
```diff
@@ -20,18 +24,57 @@ var RunCompletionStatuses = struct {
...
+		return fmt.Errorf("key: %s is nil", key)
+	} else if nn.Name == "" {
+		return fmt.Errorf("key: %s, Name field is missing", key)
+	}
+
+	return nil
+}
+
+func (sre RunCompletionEvent) Validate() error {
+	validate := validator.New()
```
Comment (grahamia) [inline argo/common/run_completion.go:54]: This shouldnt be done here. validator.New() is thread safe (well according to the comments in the code) so should only be created once and used not created on each message

**Review by grahamia — CHANGES_REQUESTED** [review]

File: argo/common/run_completion_event_unit_test.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+import (
+	"fmt"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Context("RunCompletionEvent.Validate", func() {
+
+	validate, _ = InitialiseValidation()
```
Comment (grahamia) [inline argo/common/run_completion_event_unit_test.go:13]: should check the error here else if it does happen to fail all you will get in the test output is probably a panic

File: argo/common/run_completion_event_unit_test.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+	namespaceName := NamespacedName{
+		Name:      "Name",
+		Namespace: "Namespace",
+	}
+
+	baseEvent := RunCompletionEvent{
```
Comment (grahamia) [inline argo/common/run_completion_event_unit_test.go:20]: It is weird to me that the baseEvent is an invalid one... usually that would be the valid one and then you make it invalid in those tests?

File: argo/common/run_completion_event_unit_test.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+		RunConfigurationName: &namespaceName,
+		RunName:              &namespaceName,
+		RunId:                "RunId",
+		Provider:             "Provider",
+	}
+
+	It("is valid when RunConfigurationName is missing and RunName is provided", func() {
+		invalidEvent := baseEvent
+		invalidEvent.RunConfigurationName = nil
+		Expect(invalidEvent.Validate(validate)).To(Succeed())
```
Comment (grahamia) [inline argo/common/run_completion_event_unit_test.go:32]: because reading this it says an invalid event is valid 🤯

Comment (TobyPinfold) [reply argo/common/run_completion_event_unit_test.go:32]: yup needs to change

File: argo/common/run_completion_event_unit_test.go (hunk 7 above)
Comment (TobyPinfold) [reply argo/common/run_completion_event_unit_test.go:20]: its because they are both set, we could make them both empty instead, which we will do now

File: argo/common/run_completion.go (hunk 5 above)
Comment (TobyPinfold) [reply argo/common/run_completion.go]: Changed in [5bc6a9f](https://github.com/sky-uk/kfp-operator/pull/352/commits/5bc6a9f67412272114e686d4ac9d403c0dbb9ccb) which uses Mutex to control the creation of the validator and registering of custom validation rules

File: argo/common/run_completion_event_unit_test.go (hunk 7 above)
Comment (TobyPinfold) [reply argo/common/run_completion_event_unit_test.go:23]: changed in [e78e58f](https://github.com/sky-uk/kfp-operator/pull/352/commits/e78e58f3ef3c6464dd3d9c2fcce26544d39f5512)

File: argo/common/run_completion_event_unit_test.go (hunk 6 above)
Comment (TobyPinfold) [reply argo/common/run_completion_event_unit_test.go]: changed in [e78e58f](https://github.com/sky-uk/kfp-operator/pull/352/commits/e78e58f3ef3c6464dd3d9c2fcce26544d39f5512)

File: argo/common/run_completion.go [hunk 9]
```diff
@@ -19,19 +24,73 @@ var RunCompletionStatuses = struct {
...
+	if nn == nil {
+		return fmt.Errorf("key: %s is nil", key)
+	} else if nn.Name == "" {
+		return fmt.Errorf("key: %s, Name field is missing", key)
+	}
+
+	return nil
+}
+
+func (sre RunCompletionEvent) Validate(validate *validator.Validate) error {
```
Comment (jmendesky) [inline argo/common/run_completion.go:72]: shouldn't this be a struct-level validator instead?

File: argo/common/common_types_unit_test.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+			serialised, err := json.Marshal(namespacedName)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(string(serialised)).To(Equal(`"` + namespace + "/" + name + `"`))
+			deserialised := NamespacedName{}
+			Expect(json.Unmarshal(serialised, &deserialised)).To(Succeed())
+			Expect(deserialised).To(Equal(namespacedName))
+		})
+	})
+
+	When("pointer is provider", func() {
```
Comment (jmendesky) [inline argo/common/common_types_unit_test.go:28]: Do you mean
```suggestion
	When("pointer is provided", func() {
```
?

File: argo/common/common_types_unit_test.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+package common
+
+import (
+	"encoding/json"
+	
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Context("Marshal NamespacedName", Serial, func() {
```
Comment (jmendesky) [inline argo/common/common_types_unit_test.go:12]: These seem unrelated? If we added them here for completeness we should say so in the PR description

File: argo/common/run_completion_event_unit_test.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+			invalidNamespacedName := namespacedName
+			invalidNamespacedName.Name = ""
+			invalidNamespacedName.Namespace = ""
+			Expect(validateNamespacedName(&invalidNamespacedName, "")).To(Not(Succeed()))
+		})
+
+	})
+})
+
+var _ = Context("RunCompletionEvent.String", func() {
```
Comment (jmendesky) [inline argo/common/run_completion_event_unit_test.go:121]: This also seems unrelated to this PR

File: argo/common/common_types_unit_test.go (hunk 11 above)
Comment (grahamia) [reply argo/common/common_types_unit_test.go:12]: Sorry, this was from my suggestion to refactor the existing tests that were all in suite_unit_test and to follow the same pattern as everywhere else and have own test files

File: argo/common/common_types_unit_test.go (hunk 10 above)
Comment (grahamia) [reply argo/common/common_types_unit_test.go:28]: See https://github.com/sky-uk/kfp-operator/pull/352/files/e78e58f3ef3c6464dd3d9c2fcce26544d39f5512#diff-351efeff6045b8ad01ff61963051f1b236fe93ba9e21243be970e12f79bf7003L34 🐒

**Review by jmendesky — COMMENTED** [review]
Can we please link to the newly introduced library? It makes it easier for reviewers (and future reference) to understand.

File: argo/common/common_types_unit_test.go (hunk 11 above)
Comment (grahamia) [reply argo/common/common_types_unit_test.go:12]: I have added a note to the description. 👍🏼

File: argo/common/common_types_unit_test.go (hunk 10 above)
Comment (jmendesky) [reply argo/common/common_types_unit_test.go:28]: Hahaha OK 😆 
still 🧐

File: argo/common/run_completion.go [hunk 13]
```diff
@@ -19,19 +24,73 @@ var RunCompletionStatuses = struct {
...
 }
 
+var (
+	validate *validator.Validate
+	mutex    sync.Mutex
+)
+
+func InitialiseValidation() (*validator.Validate, error) {
+	if validate == nil {
+		mutex.Lock()
```
[the remark this answers is not in the document]
Comment (edinekj) [reply argo/common/run_completion.go]: We stoped using `mutex` and start using `init function` ab2e384ccef689f93479e1463161e1622c828cb4

**Comment by TobyPinfold** [comment]
This logic will no longer be used within here, the onus for validation will be on clients reading the `RunCompletionEvents` to perform validation on events to their own requirements. So closing this change.
